### PR TITLE
feat(oauth): implement oidc contract with a generic provider and preset based on google and stackit

### LINKED ISSUE #10: OIDC/OAuth2 Integration for SSO
### Area

Security / access control

### Problem or motivation

Enterprise environments use centralized identity providers (Okta, Auth0, Keycloak, Azure AD). Managing separate credentials for Tellstone adds operational overhead and diverges from existing identity workflows. Teams expect to authenticate against their existing SSO rather than managing another set of credentials.

### Proposed solution

Add a pluggable OIDC/OAuth2 provider interface that works alongside password-based auth. When configured, both password and token auth are accepted.

**Configuration:**

| Flag | Env | Description |
|------|-----|-------------|
| `--oidc-issuer` | `TSD_OIDC_ISSUER` | OIDC provider URL (e.g. `https://accounts.google.com`) |
| `--oidc-client-id` | `TSD_OIDC_CLIENT_ID` | OAuth2 client ID |
| `--oidc-client-secret` | `TSD_OIDC_CLIENT_SECRET` | OAuth2 client secret |
| `--oidc-audience` | `TSD_OIDC_AUDIENCE` | Expected audience claim |

**Design direction:**

- Pluggable `AuthProvider` interface so additional providers (JWT validation, API keys, custom tokens) can be added without modifying core code
- Token validation happens on connection establishment, not per request — zero hot-path overhead
- Cache validated tokens with short TTL to avoid redundant OIDC provider calls
- Works with both binary protocol #8 and RESP #7 AUTH flows — the AUTH handshake accepts tokens in addition to passwords
- When `--oidc-issuer` is not set, zero overhead (no interface dispatch, no token parsing)

### Alternatives considered

_No response_

### Additional context

**Reference:** [OpenID Connect Core 1.0](https://openid.net/specs/openid-connect-core-1_0.html)

### LINKED ISSUE #34: Oauth Claim Mapping for RBAC System
### Area

Security / access control

### Problem or motivation

Tellstone's RBAC policy file (`internal/rbac/config.go:33-45`) can only express
statically enumerated users (`users:` with bcrypt passwords) and role definitions.
With OIDC connection auth (#10), identities arrive as validated JWT *claims*, not
password entries — and there is no way to map those claims to roles without creating
a per-identity user record, which defeats SSO entirely.

A dangerous default lurks in a naive fallback: granting `reader` (or any role) to an
unmatched token is wrong for a database holding sensitive data. The safe default is
**fail closed** — no matching rule means no role, and no role means no access.

### Proposed solution

Add an `oauth` section to the RBAC policy schema (`internal/rbac/config.go`) that maps
token claims to roles as a pure function — no identity/user bookkeeping involved.
```yaml
roles:
  - name: admin
    rules: ["+@all"]
  - name: reader
    rules: ["+@read", "~*"]

oauth:
  rules:                    # ordered, first match wins
    - claims:               # all predicates must match (AND)
        iss: "https://iam.acess.example"
        email: "*@saxy.dev"
      role: admin
    - claims:
        iss: "https://iam.acess.example"
        groups: ["dev-*", "qa-*"]
      role: reader
  default_role: ""          # optional; empty = fail closed
```

Semantics:
- **First match wins**, evaluated top-down; unmatched claims resolve to `default_role`,
  and an empty/unset `default_role` means **no role → connection denied**. No implicit
  read-only fallback: even `reader` is a wrong default for sensitive data.
- **Glob matching** on claim values: exact, `*`-prefix (suffix match), `*-suffix`
  (prefix match); no regex. A missing claim never matches.
- **Rule validation at load**, mirroring existing strictness (`config.go:93-96`):
  unknown target role, duplicate rule, or empty rule list is a load error — a bad file
  can never half-apply.
- **Hot reload for free**: the `oauth` block rides the existing `LoadFile`/`Parse`
  path (`config.go:49-65`), so SIGHUP re-evaluates rules without restart — matching the
  RBAC hot-swap principle (#16).

Changes:
- Extend `fileConfig` in `internal/rbac/config.go` with the `oauth` block and its
  strict validation.
- Add `ResolveRole(claims map[string][]string) (*Role, bool)` in `internal/rbac` — pure,
  allocation-light (connection time, not the per-command hot path), no identity lookup.
- Wire it into the OIDC connection-auth story (#10): after signature/`iss`/`exp`
  verification, feed the token claims to `ResolveRole`.

Acceptance criteria:
- Rule ordering respected; first match wins.
- No `default_role`: an authentic-but-unmatched token is denied, never downgraded.
- Unknown target role, duplicate rule, empty rules, and a rule with an undefined
  glob → file rejected at load.
- SIGHUP with a changed mapping takes effect without restart.
- `task check` passes.

### Alternatives considered

- Fallback to `default_role: reader` — rejected: silent over-granting on sensitive data.
- `users:` entries with OAuth subjects — rejected: defeats SSO, no group mapping.
- Flat single-claim map (`"@saxy.io": admin`) — rejected: matches on one claim alone and
  can't pin `iss`/`groups`, risking privilege escalation on a typo'd claim name.

### Additional context

- RBAC policy schema + validation: `internal/rbac/config.go`
- Role store / session wiring: `internal/rbac/policy.go`, `internal/rbac/rbac.go`
- OIDC provider seam: `internal/oauth/oauth.go` (empty stub), issue #10
- Composes with: #16 (hot-swap policy), #26/#27 (RBAC core)

### PR DESCRIPTION
**Pull request by Saxy**
## Description

**Component:** 
Authentication

**Type of Change:**
- [x] New feature (non-breaking change which adds functionality)

---

## Related Issue
Closes #10 
Closes #34 

---

## Technical Deep Dive & Context

Connection-time **OIDC bearer-token authentication**: a client presents a Google/STACKIT/any-issuer `id_token` (signed JWT) as its `AUTH` credential; its claims are mapped to an RBAC role via the policy's `oauth.rules` and pinned to the connection exactly like a password session.

**Provider contract** (`internal/oauth/oauth.go`): `Provider.Verify(ctx, token) (Claims, error)` with `Claims = map[string][]string` (JWT claims may repeat or be lists; never flattened). Two error classes — `ErrInvalidToken` for bad credentials (callers find it via `errors.Is`), unwrapped transient errors so "IdP down" is distinguishable from "bad credential". No secrets in `Config`: verification is signature + issuer + audience against the public JWKS, so **no client secret is needed server-side**.

**Generic OIDC provider** (`internal/oauth/generic`, stdlib-only, fail-fast): synchronous discovery + JWKS fetch in `New` (wrong issuer/IdP fails at startup, not first `AUTH`); algorithm allowlist RS256/ES256 only (HS* rejected to prevent algorithm-confusion); `kid` required; exactly one JWKS refresh on unknown `kid` (rotation without restarts); `exp`/`nbf`/`iss`/`aud` enforced; JWKS cache behind an `RWMutex`, whole-set replacement on refresh. Presets `google`/`stackit` are thin wrappers — Google maps the `hd` claim into `groups` so `claim: groups, match: "*@example.com"` works one-line. Registry (`internal/oauth/registry.go`) is first-wins, read-only after startup, lock-free reads.

**RBAC integration** (`internal/rbac/policy.go`): `oauth.rules` compiled at policy load (unknown target role fails the load; a middle `*` wildcard is rejected). `Store.ResolveOAuthToken` is a closure so `rbac` stays independent of the `oauth` package. First-match-wins; no match or verify failure = no role = deny-all (**fail-closed**). Session subject = token `sub` (falls back to `"default"`); audit events and `ACL LOG` account failures identically to the password path. SIGHUP hot-reload never re-evaluates pinned sessions.

**Hot-path allocation (binary client, `client.go` / `client_role.go`)**: real `id_token`s routinely exceed the 512-byte stack buffer used by `Auth`/`AuthUser`. The stack path stays allocation-free; payloads that overflow fall back to a one-time heap buffer. Before this, >512-byte credentials were rejected with `ErrRequestTooLarge` — the bundled client could not authenticate over the binary protocol at all.

**Startup config** (`config/config.go`, `server/server.go`): `--oauth-provider` (google|stackit, empty → generic), `--oauth-issuer`, `--oauth-client-id`, each with a `TSD_*` env fallback. No flags → provider `nil` → zero behavior change and zero overhead. Token auth without `--rbac-config` is a startup error (a token can only reach a role through `oauth.rules`).

**Example** (`cmd/example/oauth`): a client that presents an `id_token` (from `-token`/`-token-file`/`TSD_ID_TOKEN`/stdin), AUTHs, issues SET/GET, and proves a forged token is denied.

---

## Performance & Benchmarks

**Workload:** N/A — no measurable change to the data path.

---

## How Has This Been Tested?

- `go test ./...` — all packages pass (`internal/oauth`, `internal/oauth/generic`, `internal/oauth/presets`, `internal/rbac`, `internal/resp` oauth suites included).
- `go vet ./...` — clean.
- Live end-to-end against the real Google issuer:
  - **RESP**: `AUTH <google id_token>` → `+OK`; SET/GET succeed; forged token → `-ERR invalid password`; session reset to `-NOAUTH` after failed AUTH.
  - **Binary protocol** (raw frame + packaged client): pre-auth GET → `INVALID_AUTH`; AUTH → `MsgAuthOk`; SET/GET OK; forged token denied; large-token path exercised by `TestClientOAuthLargeToken` (700-byte JWT-shaped credential through `client.Auth`).
- Edge cases covered by unit tests: unknown/absent `kid` → JWKS refresh once then reject; missing `kid` rejected; HS256 token rejected; expired/not-yet-valid tokens rejected; multi-value `aud` accepted; `exp` absent allowed; policy with `*@domain` prefix-glob, `*@domain` suffix-glob, exact, and bare `*` matches; first-match-wins ordering; fail-closed when no rule matches; middle-`*` pattern rejected at load; unknown target role fails load; `--oauth-provider` unknown name → startup error; token auth without `--rbac-config` → startup error.
- Manual: `cmd/example/oauth` against a live server with real Google token (AUTH OK, SET OK, GET value, forged denied `INVALID_AUTH`).

---

## Checklist

- [x] My code follows the existing code style of this project
- [x] I have added tests that prove my fix/feature works
- [x] New and existing tests pass locally (`go test ./...` and `go test -race ./...`)
- [x] I have updated the documentation (README, comments, or any relevant docs)
- [x] My changes generate no new `go vet` warnings
- [x] Any breaking changes are documented and communicated

## New Features
- Added optional OAuth/OIDC bearer-token authentication for binary and RESP connections.
- Added Google, STACKIT, and generic OIDC provider support with secure token verification.
- Added configurable claim-to-role mapping with fail-closed authorization.
- Added OAuth configuration options and an end-to-end authentication example.

## Bug Fixes
- Improved handling of large authentication credentials, safely rejecting values beyond the protocol limit.

## Documentation
- Added OAuth/OIDC setup, provider, policy, and usage documentation.
- Marked audit logging as completed in the roadmap.

### CODE REVIEW

File: cmd/example/oauth/policy.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    rules: ["+get", "~*"]
+users:
+  - name: admin
+    role: admin
+    password: "$2a$10$pcaKkTfRy.KSdNUgKszYYedE7L32P9fSEG3x1phq0EbjeYkn5WpEi"
+oauth:
+  rules:
+    - claim: email
+      match: "*@example.com"
+      role: admin
```
[the remark this answers is not in the document]
Comment (Saxy) [reply cmd/example/oauth/policy.yaml:30]: any domain can be used

File: server/server.go [hunk 2]
```diff
@@ -208,6 +220,54 @@ func (s *Server) initRBAC() error {
...
+// would silently deny every connection.
+func (s *Server) initOAuth() error {
+	cfg := s.app.GetConfig()
+	logger := s.app.GetLogger()
+	issuer := cfg.GetOAuthIssuer()
+	ocfg := oauth.Config{Issuer: issuer, ClientID: cfg.GetOAuthClientID()}
+	var (
+		p   oauth.Provider
+		err error
+	)
```
[the remark this answers is not in the document]
Comment (Saxy) [reply server/server.go:246]: done

File: server/server.go [hunk 3]
```diff
@@ -208,6 +220,54 @@ func (s *Server) initRBAC() error {
...
+		p, err = generic.New(ocfg, logger)
+	default:
+		return fmt.Errorf("unknown --oauth-provider %q (want google, stackit, or empty for generic OIDC)", cfg.GetOAuthProvider())
+	}
+	if err != nil {
+		return err
+	}
+	if s.policy == nil {
+		return errors.New("--oauth-provider requires --rbac-config: tokens map to roles via the policy's oauth.rules")
+	}
```
[the remark this answers is not in the document]
Comment (Saxy) [reply server/server.go]: done

File: internal/oauth/presets/google.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+func (g *google) Verify(ctx context.Context, token []byte) (oauth.Claims, error) {
+	claims, err := g.inner.Verify(ctx, token)
+	if err != nil {
+		return nil, err
+	}
+	if hd, ok := claims["hd"]; ok && len(hd) == 1 {
+		claims["groups"] = []string{hd[0]}
+	}
+	return claims, nil
+}
```
[the remark this answers is not in the document]
Comment (Saxy) [reply internal/oauth/presets/google.go:57]: done

File: internal/oauth/generic/generic.go [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+	if cfg.Issuer != "" {
+		if iss, ok := raw["iss"].(string); !ok || iss != cfg.Issuer {
+			return errors.New("generic: issuer mismatch")
+		}
+	}
+	if cfg.ClientID != "" && !audContains(raw["aud"], cfg.ClientID) {
+		return errors.New("generic: audience mismatch")
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (Saxy) [reply internal/oauth/generic/generic.go:393]: ok check if the exp is available 
ok && now => must be true and correct timestamp -- so the current check is way more precise

won't implement!

File: internal/oauth/presets/google_test.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+				"issuer":   srv.URL,
+				"jwks_uri": srv.URL + "/jwks",
+			})
+		case "/jwks":
+			e := big.NewInt(int64(priv.PublicKey.E))
+			_ = json.NewEncoder(w).Encode(map[string]any{"keys": []map[string]string{{
+				"kty": "RSA", "kid": kid,
+				"n": b64std(priv.PublicKey.N.Bytes()),
+				"e": b64std(e.Bytes()),
+			}}})
```
[the remark this answers is not in the document]
Comment (Saxy) [reply internal/oauth/presets/google_test.go]: done

File: internal/network/server.go [hunk 7]
```diff
@@ -592,6 +609,17 @@ func (s *Server) handleAuthMessage(c gnet.Conn, st *connState, value []byte) aut
...
+	// oauth provider (which maps claims to a role) before any username lookup.
+	if s.oauth != nil && oauth.IsJWT(password) {
+		token := make([]byte, len(password))
+		copy(token, password)
+		if s.dispatchOAuth(c, token) {
+			st.authPending = true
+			return authResult{dispatched: true}
+		}
+		return authResult{respPayload: ResponseAuthErr, respType: MsgAuthErr}
+	}
```
[the remark this answers is not in the document]
Comment (Saxy) [reply internal/network/server.go:630]: done

File: cmd/example/oauth/main.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+		return strings.TrimSpace(string(b)), nil
+	}
+	if env := os.Getenv("TSD_ID_TOKEN"); env != "" {
+		return env, nil
+	}
+	scanner := bufio.NewScanner(os.Stdin)
+	if scanner.Scan() && strings.TrimSpace(scanner.Text()) != "" {
+		return strings.TrimSpace(scanner.Text()), nil
+	}
+	return "", os.ErrNotExist
```
[the remark this answers is not in the document]
Comment (Saxy) [reply cmd/example/oauth/main.go:105]: its just an simple example

File: internal/oauth/generic/generic_test.go [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+	case "RS256":
+		sig, err = rsa.SignPKCS1v15(rand.Reader, k.priv.(*rsa.PrivateKey), crypto.SHA256, digest[:])
+	case "ES256":
+		sig, err = k.priv.(*ecdsa.PrivateKey).Sign(rand.Reader, digest[:], crypto.SHA256)
+	}
+	if err != nil {
+		panic(err)
+	}
+	return signed + "." + b64enc(sig)
+}
```
[the remark this answers is not in the document]
Comment (Saxy) [reply internal/oauth/generic/generic_test.go:153]: done
